File: Realize/3Week/PracticeDeque.cpp
#include <iostream>
using namespace std;

class MyDeque2 {
    int* arr;
    int front;
    int rear;
    int size;

public:
    MyDeque2(int size) {
        this->size = size + 1;
        this->arr = new int[this->size];
        this->front = 0;
        this->rear = 0;
    }

    ~MyDeque2() {
        delete[] arr;
    }

    bool isEmpty() {
        return this->rear == this->front;
    }

    bool isFull() {
        return (this->rear + 1) % this->size == this->front;
    }

    void addFirst(int data) {
        if (this->isFull()) {
            cout << "Deque is full!" << endl;
            return;
        }
        this->front = (this->front - 1 + this->size) % this->size;
        this->arr[this->front] = data;
    }

    void addLast(int data) {
        if (this->isFull()) {
            cout << "Deque is full!" << endl;
            return;
        }
        this->arr[this->rear] = data;
        this->rear = (this->rear + 1) % this->size;
    }

    int removeFirst() {
        if (this->isEmpty()) {
            cout << "Deque is empty!" << endl;
            return -1;
        }
        int data = this->arr[this->front];
        this->front = (this->front + 1) % this->size;
        return data;
    }

    int removeLast() {
        if (this->isEmpty()) {
            cout << "Deque is empty!" << endl;
            return -1;
        }
        this->rear = (this->rear - 1 + this->size) % this->size;
        return this->arr[this->rear];
    }

    void printDeque() {
        if (this->isEmpty()) {
            cout << "Deque is empty!" << endl;
            return;
        }
        int i = this->front;
        while (i != this->rear) {
            cout << this->arr[i] << " ";
            i = (i + 1) % this->size;
        }
        cout << endl;
    }
};

int main() {
    MyDeque2 myDeque(5);

    myDeque.addFirst(1);
    myDeque.addFirst(2);
    myDeque.addFirst(3);
    myDeque.printDeque();  // 3 2 1

    myDeque.addLast(10);
    myDeque.addLast(20);
    myDeque.addLast(30);  // Deque is full!
    myDeque.printDeque();  // 3 2 1 10 20

    cout << myDeque.removeFirst() << endl; // 3
    myDeque.printDeque();  // 2 1 10 20

    cout << myDeque.removeLast() << endl; // 20
    myDeque.printDeque();  // 2 1 10

    cout << myDeque.removeLast() << endl; // 10
    cout << myDeque.removeLast() << endl; // 1
    cout << myDeque.removeLast() << endl; // 2
    cout << myDeque.removeLast() << endl; // Deque is empty!

    return 0;
}

File: Realize/3Week/PracticeDeque.py
class MyDeque2:
    def __init__(self, size):
        self.size = size + 1
        self.arr = [0] * self.size
        self.front = 0
        self.rear = 0

    def is_empty(self):
        return self.rear == self.front

    def is_full(self):
        return (self.rear + 1) % self.size == self.front

    def add_first(self, data):
        if self.is_full():
            print("Deque is full!")
            return
        self.front = (self.front - 1 + self.size) % self.size
        self.arr[self.front] = data

    def add_last(self, data):
        if self.is_full():
            print("Deque is full!")
            return
        self.arr[self.rear] = data
        self.rear = (self.rear + 1) % self.size

    def remove_first(self):
        if self.is_empty():
            print("Deque is empty!")
            return -1
        data = self.arr[self.front]
        self.front = (self.front + 1) % self.size
        return data

    def remove_last(self):
        if self.is_empty():
            print("Deque is empty!")
            return -1
        self.rear = (self.rear - 1 + self.size) % self.size
        return self.arr[self.rear]

    def print_deque(self):
        if self.is_empty():
            print("Deque is empty!")
            return
        i = self.front
        while i != self.rear:
            print(self.arr[i], end=' ')
            i = (i + 1) % self.size
        print()

if __name__ == "__main__":
    my_deque = MyDeque2(5)
    
    my_deque.add_first(1)
    my_deque.add_first(2)
    my_deque.add_first(3)
    my_deque.print_deque()  # 3 2 1

    my_deque.add_last(10)
    my_deque.add_last(20)
    my_deque.add_last(30)  # Deque is full!
    my_deque.print_deque()  # 3 2 1 10 20

    print(my_deque.remove_first())  # 3
    my_deque.print_deque()  # 2 1 10 20

    print(my_deque.remove_last())  # 20
    my_deque.print_deque()  # 2 1 10

    print(my_deque.remove_last())  # 10
    print(my_deque.remove_last())  # 1
    print(my_deque.remove_last())  # 2
    print(my_deque.remove_last())  # Deque is empty!
